python/tests/xir/test_integration.py

```diff
@@ -14,10 +14,10 @@ def parse_script(circuit: str) -> XIRProgram:
 
 
 photonics_script = """
-gate Sgate, 2, 1;
-gate BSgate, 2, 2;
-gate Rgate, 1, 1;
-output MeasureHomodyne;
+gate Sgate(a, b)[0];
```

Comment:
Is it possible to declare gates without specifying the wires they act on, and to later on then use the gate and at the same time specify the wires?

Reply:
I don't think so but I'll let @thisac confirm. In general, I think we should enforce a declare-before-you-use rule in the validator.

Reply:
Thanks! So is the following the preferred style for declaration and execution?
```python
gate BSgate(theta, phi)[0, 1];

BSgate(0.1, 0.0) | [0, 1];
```
To me this seems a little too much wire info (e.g., `[0, 1]` is twice) versus:
```python
gate BSgate(theta, phi);

BSgate(0.1, 0.0) | [0, 1];
```
Though, looking at [ir.lark](https://github.com/XanaduAI/jet/blob/498b95136a3c5273082650b290e08b570f6b93b1/python/xir/ir.lark#L12) perhaps the wires in a gate definition are optional? :thinking: Ah, indeed, maybe [this](https://github.com/XanaduAI/jet/blob/498b95136a3c5273082650b290e08b570f6b93b1/python/tests/xir/test_integration.py#L79) is an example? Perhaps @thisac you can share your thoughts.

I'm also curious what happens if you do:
```python
gate H2:
    H | [0];
    H | [1];
end;

H2 | [0, 2];
```
Does this raise an error? Are Hadamards applied to `[0, 1]` or `[0, 2]`?

Reply:
Good questions, @trbromley!

> Thanks! So is the following the preferred style for declaration and execution?
> 
> ```python
> gate BSgate(theta, phi)[0, 1];
> 
> BSgate(0.1, 0.0) | [0, 1];
> ```

Yes, although you would also need to define the gate too (assuming it's not provided by the interpreter).

---

> To me this seems a little too much wire info (e.g., `[0, 1]` is twice) versus:
> 
> ```python
> gate BSgate(theta, phi);
> 
> BSgate(0.1, 0.0) | [0, 1];
> ```

True, but the problem now is that it is unclear how many wires `BSgate` acts upon:
```haskell
gate BSgate(theta, phi);
```

Note that the wire labels/aliases to be used in the `BSgate` definition are also unknown (although this problem is more of an XIR interpreter implementation detail than anything else).

---

> Though, looking at [ir.lark](https://github.com/XanaduAI/jet/blob/498b95136a3c5273082650b290e08b570f6b93b1/python/xir/ir.lark#L12) perhaps the wires in a gate definition are optional? 🤔 Ah, indeed, maybe [this](https://github.com/XanaduAI/jet/blob/498b95136a3c5273082650b290e08b570f6b93b1/python/tests/xir/test_integration.py#L79) is an example?  Perhaps @thisac you can share your thoughts.

Ah, the wire labels in a gate *definition* are optional but the wire labels in a gate *declaration* are **not** optional.  In the former case, the wire labels are implicitly assumed to be `[0, 1, ..., n - 1]` where `n` (I believe) is the number of unique wire labels in the definition of the gate.  Of course, this information is not accessible in the gate declaration.

---

> 
> I'm also curious what happens if you do:
> 
> ```python
> gate H2:
>     H | [0];
>     H | [1];
> end;
> 
> H2 | [0, 2];
> ```
> 
> Does this raise an error? Are Hadamards applied to `[0, 1]` or `[0, 2]`?

Nope, this is well-defined!  The program you wrote is equivalent to
```haskell
gate H2 | [0, 1]:
    H | [0];
    H | [1];
end;
 
H2 | [0, 2];
```
where `0` is aliased to `0` and `1` is aliased to `2` inside the gate definition.  So, `H2` is applied to wires `0` and `2`.

Reply:
Thanks @Mandrenkov this is very useful!

Are gate declarations only needed when a gate definition is not provided? E.g., declarations are for device/simulator supported gates?

Reply:
> Thanks @Mandrenkov this is very useful!

No problem!

> Are gate declarations only needed when a gate definition is not provided? E.g., declarations are for device/simulator supported gates?

Currently, we expect _all_ gates to be declared beforehand.  The only difference between a user-specified gate (e.g., `H2`) and an implementation-provided gate (e.g., `H`) is that a user-specified gate must also have an associated gate definition (the gate definition of the implementation-provided gate is implicit and effectively atomic).

As you pointed out though, this introduces some redundancy.  One thing we _could_ do is have each gate definition implicitly create a gate declaration if one does not already exist.  In other words,
```haskell
gate RotXYZ(theta) | [0]:
    RX(theta) | [0];
    RY(theta) | [0];
    RZ(theta) | [0];
end;
```
would be shorthand for
```haskell
gate RotXYZ(theta) | [0];

gate RotXYZ(theta) | [0]:
    RX(theta) | [0];
    RY(theta) | [0];
    RZ(theta) | [0];
end;
```
Any thoughts on this, @thisac?

Reply:
This is a great idea!

Reply:
Sorry for being a bit late to the discussion. So, I would've said that a gate definition would be enough (i.e. a declaration wouldn't be necessary in that case). So when checking statements for valid gates, both the declarations and the custom gate definitions would be checked. 

Would we actually need to add a declaration for those as well (implicitly or explicitly)? If so, I'd also be in favour of adding a declaration implicitly for each gate definition, which could easily be done in the transformation stage.

I also think that it might be good to have wires be optional for gate declarations. Specifically, as an example, there might be gate (such as the quantum Fourier transform, `gate QFT;`) that have no set number of wires. I like the idea of having that flexibility, at least in the IR. It would also be nice to simply be able to declare the _number_ of wires (and implicitly labelling them 0, 1, 2, etc.), although I'm not sure which syntax we should use in that case.

> Though, looking at ir.lark perhaps the wires in a gate definition are optional? 🤔 Ah, indeed, maybe this is an example? Perhaps @thisac you can share your thoughts.

The case here is that, as @Mandrenkov pointed out, the wires are implicitly extracted from the statements if not declared. This wouldn't be the case for the new declarations in this PR, since they have no statements to extract the wires from.

Reply:
It's possible to use the range syntax for wires:
```python
gate BSgate(theta, phi)[0..2]
```
An idea would be to change the range to interpret e.g. `[..2]` as `[0, 1]` providing a shorthand for number of wires:
```python
gate BSgate(theta, phi)[..2]
gate Toffoli[..3]
gate BigU[..15]
```

Reply:
I like the range syntax idea! It gives flexibility to explicitly declare with `[0, 1, 2]` or use a range `[..3]` :+1: Nice idea!

Reply:
> That's a good point! I'm not sure that we need this degree of flexibility in v0.1.0 of the XIR, but when it does become a priority, I would suggest using ellipses:

+1 for the ellipses! And agree that this may be out of scope for v0.1.0. Suppose we attempt to run a PL circuit which includes a QFT gate, our plan for v0.1.0 would be to raise an error before serialization?

Reply:
> Suppose we attempt to run a PL circuit which includes a QFT gate, our plan for v0.1.0 would be to raise an error before serialization?

Great question!  I think we have two options:
1. Raise an Exception during serialization (as you mentioned), or
2. Declare each used "instance" of the QFT gate (e.g., `gate QFT_16 [0..15]`).

Reply:
I also like the ellipses! 👍 

My only concern at the moment is that always having to declare wire labels can be confusing. As with the case
```python
gate BSgate(theta, phi)[0, 1];
BSgate(0.1, 0.0) | [0, 2];
```
Where, in the declaration, 0 and 1 are labels for only that specific gate. It's clearer if using string labels.
```python
gate BSgate(theta, phi)[a, b];
BSgate(0.1, 0.0) | [0, 2];
```
but I find these names quite arbitrary. Perhaps just using the range, e.g. `gate BSgate(theta, phi)[0..2];`, as standard syntax for a number of wires could work, although I'd prefer another way to just say `gate BSgate(theta, phi)` on 2 wires (which could then use the integer labelling implicitly.

> Great question! I think we have two options:
> 
>     1. Raise an Exception during serialization (as you mentioned), or
>     2. Declare each used "instance" of the QFT gate (e.g., gate QFT_16 [0..15]).

I think both of these options are good, with a preference towards the second one.

Reply:
> My only concern at the moment is that always having to declare wire labels can be confusing. As with the case

Could this be solved by making it more explicit? For example, requiring the statement `wires`?

```python
gate BSgate(theta, phi) wires[0, 1]
```

Reply:
We should probably continue this discussion on Slack - it is easy to lose track of GitHub conversations!